packages/workflow/src/records/handler/create.ts

```diff
@@ -222,149 +240,385 @@ async function createRecord(
 
   const responseBundle = await sendBundleToHearth(inputBundle)
   const savedBundle = toSavedBundle(inputBundle, responseBundle)
-  const record = inProgress
-    ? changeState(savedBundle, 'IN_PROGRESS')
-    : changeState(savedBundle, 'READY_FOR_REVIEW')
 
-  const mergedBundle = mergeBundles(record, practitionerResourcesBundle)
+  const mergedBundle = mergeBundles(savedBundle, practitionerResourcesBundle)
 
-  const encounter = findEncounterFromRecord(record)
+  const encounter = findEncounterFromRecord(savedBundle)
   if (encounter == null) {
-    return mergedBundle
+    return mergedBundle as T
   }
 
-  if (bundleIncludesLocationResources(record)) {
-    return mergedBundle
+  if (bundleIncludesLocationResources(savedBundle)) {
+    return mergedBundle as T
   }
 
   const locationResourcesBundle = await resolveLocationsForEncounter(encounter)
 
   if (locationResourcesBundle == null) {
-    return mergedBundle
+    return mergedBundle as T
   }
 
-  return mergeBundles(mergedBundle, locationResourcesBundle)
+  return mergeBundles(mergedBundle, locationResourcesBundle) as T
 }
 
-type CreatedRecord =
-  | InProgressRecord
-  | ReadyForReviewRecord
-  | ValidatedRecord
-  | WaitingForValidationRecord
-
+function getEventAction(record: ValidatedRecord): 'sent-for-approval'
+function getEventAction(
+  record: InProgressRecord | ReadyForReviewRecord
+): 'sent-notification' | 'sent-notification-for-review'
+function getEventAction(
+  record: ReadyForReviewRecord
+): 'sent-notification-for-review'
+function getEventAction(
+  record: WaitingForValidationRecord
+): 'waiting-external-validation'
 function getEventAction(record: CreatedRecord) {
   if (isInProgress(record)) {
-    return 'sent-notification'
+    return 'sent-notification' as const
   }
   if (isReadyForReview(record)) {
-    return 'sent-notification-for-review'
+    return 'sent-notification-for-review' as const
   }
   if (isValidated(record)) {
-    return 'sent-for-approval'
+    return 'sent-for-approval' as const
   }
   if (isWaitingExternalValidation(record)) {
-    return 'waiting-external-validation'
+    return 'waiting-external-validation' as const
   }
   // type assertion
   record satisfies never
   // this should never be reached
-  return 'sent-notification'
+  return 'sent-notification' as const
 }
 
-export default async function createRecordHandler(
-  request: Hapi.Request,
-  _: Hapi.ResponseToolkit
+export async function declareRecordHandler(
```

Comment:
This function and the two other handlers `validateRecordHandler` and `registerRecordHandler` follow one simple rule: you can call these functions as many times you like with the same parameters and the end result is always the same. So call this function once or million times, you should always only see one record in the database. Idempotency like this is absolutely critical so that the message queue can submit the same record here twice.

Reply:
I mean are the notification endpoints idempotent too?

Reply:
Not yet. I left them untouched for now as in many cases it's the last operation to be performed. We should make them idempotent though. I'll add it to the TODO list